src/mame/amstrad/pda600_copro.cpp

```diff
@@ -0,0 +1,403 @@
+//    |  0x01  | Length |  Type  | Image[32] | Strokes | Flags |  Char  | Checksum |
+//    +--------+--------+--------+-----//----+---------+-------+--------+----------+
+//
+//    Image, Strokes, Flags:
+//     Same as recognition frame.
+//
+//    Char:
+//     Character to be trained with the given image.
+
+	m_resp_type = 'Y';
+	m_state = STATE_BUSY;
+	m_update_timer->adjust(attotime::from_msec(50));    // Time required to complete the training
+}
+
+
+u8 pda600_copro_device::recognize_char()
+{
+	// We don't implement any character recognition functionality, but simply return
+	// the character corresponding to the key pressed.
+
+	if (machine().input().code_pressed(KEYCODE_0)) return '0';
+	if (machine().input().code_pressed(KEYCODE_1)) return '1';
+	if (machine().input().code_pressed(KEYCODE_2)) return '2';
+	if (machine().input().code_pressed(KEYCODE_3)) return '3';
+	if (machine().input().code_pressed(KEYCODE_4)) return '4';
+	if (machine().input().code_pressed(KEYCODE_5)) return '5';
+	if (machine().input().code_pressed(KEYCODE_6)) return '6';
+	if (machine().input().code_pressed(KEYCODE_7)) return '7';
+	if (machine().input().code_pressed(KEYCODE_8)) return '8';
+	if (machine().input().code_pressed(KEYCODE_9)) return '9';
+	if (machine().input().code_pressed(KEYCODE_A)) return 'A';
+	if (machine().input().code_pressed(KEYCODE_B)) return 'B';
+	if (machine().input().code_pressed(KEYCODE_C)) return 'C';
+	if (machine().input().code_pressed(KEYCODE_D)) return 'D';
+	if (machine().input().code_pressed(KEYCODE_E)) return 'E';
+	if (machine().input().code_pressed(KEYCODE_F)) return 'F';
+	if (machine().input().code_pressed(KEYCODE_G)) return 'G';
+	if (machine().input().code_pressed(KEYCODE_H)) return 'H';
+	if (machine().input().code_pressed(KEYCODE_I)) return 'I';
+	if (machine().input().code_pressed(KEYCODE_J)) return 'J';
+	if (machine().input().code_pressed(KEYCODE_K)) return 'K';
+	if (machine().input().code_pressed(KEYCODE_L)) return 'L';
+	if (machine().input().code_pressed(KEYCODE_M)) return 'M';
+	if (machine().input().code_pressed(KEYCODE_N)) return 'N';
+	if (machine().input().code_pressed(KEYCODE_O)) return 'O';
+	if (machine().input().code_pressed(KEYCODE_P)) return 'P';
+	if (machine().input().code_pressed(KEYCODE_Q)) return 'Q';
+	if (machine().input().code_pressed(KEYCODE_R)) return 'R';
+	if (machine().input().code_pressed(KEYCODE_S)) return 'S';
+	if (machine().input().code_pressed(KEYCODE_T)) return 'T';
+	if (machine().input().code_pressed(KEYCODE_U)) return 'U';
+	if (machine().input().code_pressed(KEYCODE_V)) return 'V';
+	if (machine().input().code_pressed(KEYCODE_W)) return 'W';
+	if (machine().input().code_pressed(KEYCODE_X)) return 'X';
+	if (machine().input().code_pressed(KEYCODE_Y)) return 'Y';
+	if (machine().input().code_pressed(KEYCODE_Z)) return 'Z';
+	if (machine().input().code_pressed(KEYCODE_MINUS)) return '-';
+	if (machine().input().code_pressed(KEYCODE_STOP)) return '.';
```

Comment:
Using hard-coded inputs like this for anything besides debugging isn’t acceptable.  You’ll need to use a different approach.

Reply:
This device has no input ports, it receives a 16x16 monochrome image of the handwritten character and returns the recognized character, so I'm not sure what's the best way to proceed, I'm open to suggestions.